stdlib/public/core/HashedCollections.swift.gyb

```diff
@@ -561,13 +561,24 @@ public struct Set<Element : Hashable> :
   public func intersect<
     S : SequenceType where S.Generator.Element == Element
   >(sequence: S) -> Set<Element> {
-    let other = sequence as? Set<Element> ?? Set(sequence)
-    var newSet = Set<Element>()
-    for member in self {
-      if other.contains(member) {
-        newSet.insert(member)
+
+    // Attempt to iterate the smaller between `self` and `sequence`.
+    // If sequence is not a set, iterate over it because it may be single-pass.
+    var newSet: Set<Element>
+
+    if let other = sequence as? Set<Element> {
+      newSet = Set<Element>(minimumCapacity: min(other.count, count))
```

Comment:
@gribozavr @florianreinhart I ran the benchmark for both options and here's the result.

```
newSet = Set<Element>()
small ∩ large * 100 -- 0.313429653644562
large ∩ small * 100 -- 9.32736319303513
small ∩ small * 100 -- 0.30213850736618
large ∩ large * 100 -- 19.0307229161263
total time spent ----- 28.9736542701721

newSet = Set<Element>(minimumCapacity: min(other.count, count))
small ∩ large * 100 -- 0.262056231498718
large ∩ small * 100 -- 9.16748160123825
small ∩ small * 100 -- 0.256339490413666
large ∩ large * 100 -- 15.4998977780342
total time spent ----- 25.1857751011848
```